---
title: "cdsi Load"
author: OW_SSB
output: html_myeebook
editor_options: 
  chunk_output_type: console
---

```{r, warning=FALSE, message=FALSE, warning=FALSE, echo=F}
rm(list=ls())
ROOT_DIR<-"/fh/fast/furlan_s/experiments/Ping"
DATA_DIR <- file.path(ROOT_DIR, "data")      # SPECIFY HERE
RES_DIR  <- file.path(ROOT_DIR,  "res")     # SPECIFY HERE
RMD_DIR  <- file.path(ROOT_DIR,  "rmd")     # SPECIFY HERE
CDS_DIR <- file.path(ROOT_DIR,   "cds")
FIG_DIR <- file.path(ROOT_DIR,  "figs")

suppressPackageStartupMessages({
  library(monocle3)
  library(m3addon)
  library(dplyr)
  library(Matrix)
  library(reticulate)
  library(ggplot2)
  library(pals)
  library(RColorBrewer)
  library(Seurat)
  library(ComplexHeatmap)
  library(ArchR)
  library(Biobase)
  library(stringr)
  library(viridis)
  library(Seurat)
  library(parallel)
  library(scCustomize)
  library(scRepertoire)
  library(forcats)
  library(ggalluvial)
  library(future)
  library(pbmcapply)
  library(parallel)
  library(dplyr)
  library(msigdbr)
  library(fgsea)
  library(stringr)
  library(harmony)
  library(future)
  library(ggplot2)
  library(circlize)
  library(scales)
})

set.seed(1234)

dyn.load('/app/software/ArrayFire/3.8.1/lib64/libaf.so.3')
library(RcppArrayFire)
library(viewmastR)

plan("multisession", workers = 12)
options(future.globals.maxSize = 60000000 * 1024^2)
```
# Make object with MONOCLE
## load data
```{r}
folders<-list.files(DATA_DIR, include.dirs = T, full.names = T)
folders<-folders[1:3]      # subset out Simone's sample
#debug(load_cellranger_data_h5)
cds<-m3addon::load_cellranger_data_h5(rev(folders), samplenames = rev(basename(folders)))
ccol<-pals::glasbey(n=15)
names(ccol)<-levels(pData(cds)$sample)
prot<-cds[fData(cds)$feature_type=="Antibody Capture",]
prot<-as.matrix(t(exprs(prot)))
log_prot<-log(prot)
colnames(log_prot)<-paste0("Log_", colnames(log_prot))

cds<-cds[fData(cds)$feature_type=="Gene Expression",]
pData(cds)<-cbind(pData(cds), cbind(prot, log_prot) )
cds<-estimate_size_factors(cds)
cds<-detect_genes(cds)
cds<-calculate_gene_dispersion(cds, method = "m3addon")
```
## Metrics
```{r, echo=F, results="as.is"}
se<-lapply(folders, function(folder) read.csv(file.path(folder, "outs", "metrics_summary.csv")))
names(se)<-basename(folders)
se<-do.call(rbind, se)
knitr::kable(se)
```
## UMI per cell per sample
```{r, echo=F}
pData(cds)$n_umi<-colSums(exprs(cds))   ## pData same thing as coldata
pData(cds)$n_gene<-apply(exprs(cds), 2, function(col) {
  as.numeric(col)
  length(which(col!=0))})
pData(cds)$log_umi<-log(pData(cds)$n.umi, base=10)
qc<-data.frame(umi_per_cell=pData(cds)$n_umi, sample=pData(cds)$sample, gene_per_cell=pData(cds)$n_gene, log_umi=pData(cds)$log_umi)

ggplot(qc, aes(x=umi_per_cell, fill=sample))+
  geom_density(alpha=0.4)

ggplot(qc, aes(x=log_umi, fill=sample))+
  geom_density(alpha=0.4)
```

## Dimensionality Reduction
Let's look at a plot of log normalized mean gene expression (x axis) and log of CV^2 (y axis).  Blue dots indicate those genes used in downstream clustering as they are the most variant genes irrespective of expression level. 
```{r, echo=F}
cds<-select_genes(cds, fit_min = 1.03, logmean_ll = -6.5)
plot_gene_dispersion(cds)
cds<-preprocess_cds(cds, num_dim = 50,  verbose = T, use_genes = get_ordering_genes(cds))
```

# QC
```{r}
mito.genes <- fData(cds)$id[grep(pattern = "^mt-", x = fData(cds)$gene_short_name)]
pData(cds)$percent.mito <- Matrix::colSums(exprs(cds[mito.genes, ]))/Matrix::colSums(exprs(cds))

quantile(pData(cds)$percent.mito , c(0.05, .5, .95))

pData(cds)$likely_dead <- pData(cds)$percent.mito>0.15
```
## UMAP clustering
Now let's look at the % variance PCs for all of these genes across all cells.
```{r, echo=F}
plot_pc_variance_explained(cds)
```

## The top 20 PCs will be fed into UMAP...
```{r, echo=F, include=F}
cds<-reduce_dimension(cds, reduction_method = "UMAP", num_dim = 20, verbose=T, cores=2)
cds<-cluster_cells(cds, resolution = 3e-4, verbose=F)
```

```{r, echo=F}
plot_cells(cds, color_cells_by = "sample", label_cell_groups = F, cell_size = 0.2)+scale_color_manual(values=ccol)
```

## QC Plots on UMAP embedding
```{r}
plot_cells(cds, color_cells_by = "percent.mito", label_cell_groups = F)
plot_cells(cds, color_cells_by = "likely_dead", label_cell_groups = F)
plot_cells(cds, color_cells_by = "n.umi", label_cell_groups = F)
plot_cells(cds, color_cells_by = "log_umi", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT1", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT2", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT3", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT4", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT5", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT6", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT7", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT8", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_HT9", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_CD8", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_CD4", label_cell_groups = F)
plot_cells(cds, color_cells_by = "Log_CD45_1", label_cell_groups = F)
plot_cells(cds, genes = "Cd3e")
plot_cells(cds, genes = "Cd8a")
```
## filter on log umi and %mito
```{r, echo=F}
lt<-log10(2000)
ht<-log10(20000)
ggplot(qc, aes(x=log_umi, fill=sample))+
  geom_density(alpha=0.4)+scale_fill_manual(values=ccol)+geom_vline(xintercept = c(lt, ht))

cdsT<-cds[,pData(cds)$log_umi>lt & pData(cds)$log_umi < ht]
cdsT<-cdsT[,!pData(cdsT)$likely_dead]

plot_cells(cdsT, color_cells_by = "percent.mito", label_cell_groups = F)

cdsT<-select_genes(cdsT,  logmean_ll = -6, top_n = 2000)
cdsT<-preprocess_cds(cdsT, num_dim = 50,  verbose = T, use_genes = get_selected_genes(cdsT))
cdsT<-reduce_dimension(cdsT, reduction_method = "UMAP", num_dim = 25, verbose=T, cores=2)
```

```{r, echo=F, include=F}
cdsT<-cluster_cells(cdsT, resolution = 3e-4, verbose=T)
```

## Remove satellite clusters
```{r}
plot_cells(cdsT, color_cells_by = "partition",  label_cell_groups = F, cell_size = 0.3)
plot_cells(cdsT, genes = "Cd3e")

cds<-cdsT[,partitions(cdsT) %in% c(1)]   ## subset out the cluster of interest
plot_cells(cds, color_cells_by = "partition",  label_cell_groups = F, cell_size = 0.3)
```
## filtering cells based on hash, signal and removing doublets
```{r, echo=F}
lpd<-pData(cds)[grep("Log_", colnames(pData(cds)))] 
lpd$Cell<-rownames(lpd)
colnames(lpd)<-gsub("Log_", "", colnames(lpd))
lpd<-data.table::as.data.table(lpd)
pd<-lpd %>% tidyr::pivot_longer(-Cell, values_to = "Log_Expression", names_to = "Ab")
threshes<-data.frame(Ab=c("HT1","HT2","HT3","HT4" ,"HT5" ,"HT6" ,"HT7","HT8","HT9"), Log_Expression=c(4,4.5,4,3.5,3.5,3.5,3.5,3.5,3.5))
ggplot(pd, aes(x=Ab, y=Log_Expression, fill=Ab))+
  geom_violin(scale="width")+geom_point(data=threshes, shape=95, size=20)+ theme(legend.position = "none")

##label cells as doublets based on log HTO thresholds
cds$is_doublet<-cds$sample
df<-data.frame(pData(cds)[,grepl("Log_HT", colnames(pData(cds)))])

mat<-matrix(nrow=length(colnames(cds)), ncol = length(threshes$Ab))
rownames(mat)<-colnames(cds)
colnames(mat)<-threshes$Log_Expression
for(i in 1:length(threshes$Log_Expression)){
  mat[,colnames(mat)[i]]<-df[,colnames(df[i])]
}
mat<-mat>colnames(mat)
mat[which(is.na(mat))]<-0
mat[which(mat ==F)]<-0
mat[which(mat ==T)]<-1

cds$is_doublet<-rowSums(mat) > 1
plot_cells(cds, color_cells_by = "is_doublet")
table(cds$is_doublet)
cds<-cds[,cds$is_doublet == F]
```

# assigning metadata based on experimental design of hashes/lanes
```{r, echo=F}
cds$id[cds$sample == "PZ_1" & cds$Log_HT1 > threshes$Log_Expression[1]]<- "Saline_1"
cds$id[cds$sample == "PZ_1"  & cds$Log_HT2 > threshes$Log_Expression[2]]<- "Saline_2"
cds$id[cds$sample == "PZ_1" & cds$Log_HT3 > threshes$Log_Expression[3]]<- "Saline_3"

cds$id[cds$sample == "PZ_2" &cds$Log_HT4 > threshes$Log_Expression[4]]<- "CSA_5_1"
cds$id[cds$sample == "PZ_2" &cds$Log_HT5 > threshes$Log_Expression[5]]<- "CSA_5_2"
cds$id[cds$sample == "PZ_2" &cds$Log_HT6 > threshes$Log_Expression[6]]<- "CSA_5_3"

cds$id[cds$sample == "PZ_3" &cds$Log_HT7 > threshes$Log_Expression[7]]<- "CSA_50_1"
cds$id[cds$sample == "PZ_3" &cds$Log_HT8 > threshes$Log_Expression[8]]<- "CSA_50_2"
cds$id[cds$sample == "PZ_3" &cds$Log_HT9 > threshes$Log_Expression[9]]<- "CSA_50_3"

# cds$mouse[cds$sample == "493_3"  & cds$Log_HT1 > threshes$Log_Expression[1]]<- "M2-13"
# cds$mouse[cds$sample == "493_3"& cds$Log_HT2 > threshes$Log_Expression[2]]<- "M2-0"

cds<-cds[,!is.na(cds$id)]

table(cds$id)

cds$cond <- cds$id

cds$cond[cds$id == "Saline_1" | cds$id == "Saline_2" | cds$id == "Saline_3"]<-
"PBS_controlled"

cds$cond[cds$id == "CSA_5_1" | cds$id == "CSA_5_2" | cds$id == "CSA_5_3"]<-
"Low_dose"

cds$cond[cds$id == "CSA_50_1" | cds$id == "CSA_50_2" | cds$id == "CSA_50_3"]<-
"High_dose"

# cds$cond[cds$id == "M2-13" | cds$id == "M2-0"]<-
# "PBS_controlled"

table(cds$cond)

plot_cells(cds, color_cells_by = "cond", label_cell_groups = F, cell_size = 0.75)+scale_color_manual(values =c("#ffa305", "#8ed1e6" ,"#33A02C",  "#db096f"))+facet_wrap(~cond)      ## facet wrap to change order

levels(factor(cds$cond))  ## to check the order of the dataset

#save the object here
#saveRDS(cds, file.path(CDS_DIR, "ping_cds.rds"))

plot_cells(cds[,order(cds$cond, decreasing = T)], color_cells_by = "cond", label_cell_groups = F, cell_size = 0.4)+scale_color_manual(values =c("#ffa305",  "#33A02C", "#8ed1e6", "#db096f"))

plot_cells(cds, color_cells_by = "sample", label_cell_groups = F, cell_size = 0.4)
plot_cells(cds, color_cells_by = "cluster", label_cell_groups = F)+scale_color_manual(values = clus_cols)

plot_cells(cds, color_cells_by = "id", label_cell_groups = F)
```

# Cd4 and cd8 subset based on gex (seurat)
```{r}
seu<-monocle3_to_seurat(cds)

DefaultAssay(seu) <- 'RNA'
seu <-  FindVariableFeatures(seu) %>% ScaleData() %>% RunPCA()

#subset based on RNA counts and whether cells pass ADT thresholds!!!!!!!!!!!
seu4<-subset(seu, subset = Cd8a <= 0 & Cd4 > 0 & Log_CD4 > 4 & Log_CD8 < 3)
seu8<-subset(seu, subset = Cd4 <= 0 & Cd8a > 0 & Log_CD4 < 4 & Log_CD8 > 3)
```
# run RPCA
```{r cd4}
# split the dataset into a list of two seurat objects (stim and CTRL)
cd4.list <- SplitObject(seu4, split.by = "sample")

# normalize and identify variable features for each dataset independently
cd4.list <- lapply(X = cd4.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration run PCA on each
# dataset using these features
features <- SelectIntegrationFeatures(cd4.list)
cd4.list <- lapply(X = cd4.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})

cd4.anchors <- FindIntegrationAnchors(cd4.list, anchor.features = features, reduction = "rpca")

# this command creates an 'integrated' data assay
cd4.combined <- IntegrateData(anchorset = cd4.anchors)

# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(cd4.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
cd4.combined <- ScaleData(cd4.combined, verbose = FALSE)
cd4.combined <- RunPCA(cd4.combined, npcs = 30, verbose = FALSE)
ElbowPlot(cd4.combined)
cd4.combined <- RunUMAP(cd4.combined, reduction = "pca", dims = 1:20)
cd4.combined <- FindNeighbors(cd4.combined, reduction = "pca", dims = 1:30)
cd4.combined <- FindClusters(cd4.combined, resolution = 0.4, graph.name = "integrated_snn")
DimPlot(cd4.combined, group.by = "cond")


DefaultAssay(cd4.combined) <- "RNA"

FeaturePlot_scCustom(cd4.combined, features = c("Log_CD8", "Log_CD45_1", "Log_CD4"), colors_use = paletteContinuous(set = "whitePurple"))
FeaturePlot_scCustom(cd4.combined, features = c("Cd8a", "Cd4"), colors_use = paletteContinuous(set = "whitePurple"))

FeaturePlot_scCustom(cd4.combined, features = c("Lef1", "Sell", "Mki67", "Gzma", "Tigit", "Tox", "Il2rb" ,"Ifng", "Tcf7", "Tbx21", "Foxp3"), colors_use = paletteContinuous(n=8)[c(1:3, 6:8)])

seu4<-cd4.combined
```

```{r cd8}
# split the dataset into a list of two seurat objects (stim and CTRL)
cd8.list <- SplitObject(seu8, split.by = "sample")

# normalize and identify variable features for each dataset independently
cd8.list <- lapply(X = cd8.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration run PCA on each
# dataset using these features
features <- SelectIntegrationFeatures(cd8.list)
cd8.list <- lapply(X = cd8.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})

cd8.anchors <- FindIntegrationAnchors(cd8.list, anchor.features = features, reduction = "rpca")

# this command creates an 'integrated' data assay
cd8.combined <- IntegrateData(anchorset = cd8.anchors)

# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(cd8.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
cd8.combined <- ScaleData(cd8.combined, verbose = FALSE)
cd8.combined <- RunPCA(cd8.combined, npcs = 30, verbose = FALSE)
ElbowPlot(cd8.combined)
cd8.combined <- RunUMAP(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindNeighbors(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindClusters(cd8.combined, resolution = 0.4, graph.name = "integrated_snn", method = 2)


DimPlot(cd8.combined, group.by = "cond")
DimPlot(cd8.combined, group.by = "seurat_clusters")
DimPlot(cd8.combined, group.by = "cloneType")

DefaultAssay(cd8.combined) <- "RNA"

rownames(cd8.combined)[grep("^Tcr", rownames(cd8.combined))]

FeaturePlot_scCustom(cd8.combined, features = c("Log_CD8", "Log_CD45_1", "Log_CD4"), colors_use = paletteContinuous(set = "whitePurple"))
FeaturePlot_scCustom(cd8.combined, features = c("Cd8a", "Cd4"), colors_use = paletteContinuous(set = "whitePurple"))
FeaturePlot_scCustom(cd8.combined, features = c("percent.mito"), colors_use = paletteContinuous(set = "whitePurple"))


##filter out > 10% mito genes
cd8.combined<-cd8.combined[,cd8.combined$percent.mito < 0.05]
cd8.combined <- RunUMAP(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindNeighbors(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindClusters(cd8.combined, resolution = 0.4, graph.name = "integrated_snn", method = 2)

cd8.combined<-cd8.combined[,cd8.combined$seurat_clusters != 9]
cd8.combined <- RunUMAP(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindNeighbors(cd8.combined, reduction = "pca", dims = 1:15)
cd8.combined <- FindClusters(cd8.combined, resolution = 0.3, graph.name = "integrated_snn", method = 2)

VlnPlot(cd8.combined, features = "percent.mito", group.by = "seurat_clusters")
DimPlot(cd8.combined, group.by = "seurat_clusters")

cd8.combined<-FindSubCluster(cd8.combined, cluster = 0, graph.name = "integrated_snn", resolution = 0.2 )
DimPlot(cd8.combined, group.by = "sub.cluster")


cd8.combined$cluster1<-cd8.combined$sub.cluster
cd8.combined$cluster1[cd8.combined$sub.cluster == "0_0"]<-"8"
cd8.combined$cluster1[cd8.combined$sub.cluster == "0_1"]<-"9"
DimPlot(cd8.combined, group.by = "sub.cluster", cols = paletteDiscrete(values = levels(factor(seu8$sub.cluster)), set = "calm"))

FeaturePlot_scCustom(cd8.combined, features = c("Lef1", "Sell", "Mki67", "Gzma", "Tigit", "Tox", "Il2rb" ,"Ifng", "Tcf7", "Tbx21", "Cd226"), colors_use = paletteContinuous(n=8)[c(1:3, 6:8)])





lpd<-cd8.combined@meta.data[grep("Log_", colnames(cd8.combined@meta.data))] 
lpd$Cell<-rownames(lpd)
colnames(lpd)<-gsub("Log_", "", colnames(lpd))
lpd<-data.table::as.data.table(lpd)
pd<-lpd %>% tidyr::pivot_longer(-Cell, values_to = "Log_Expression", names_to = "Ab")
threshes<-data.frame(Ab=c("HT1","HT2","HT3","HT4" ,"HT5" ,"HT6" ,"HT7","HT8","HT9"), Log_Expression=c(4,4.5,4,3.5,3.5,3.5,3.5,3.5,3.5))
ggplot(pd, aes(x=Ab, y=Log_Expression, fill=Ab))+
  geom_violin(scale="width")+geom_point(data=threshes, shape=95, size=20)+ theme(legend.position = "none")



seu8<-cd8.combined
```

# Add TCR data
```{r}
samps<-list.files(DATA_DIR)[2:4]
# Load VDJ data (one csv per run)

bcs<-colnames(seu)

#all barcodes end in -1, the only identifier is the prefix
str_split(bcs, "\\-")%>% sapply("[[", 2) %>% table()

tcrl<-lapply(samps, function(x){
  tcr<-read.csv(file.path(DATA_DIR, x, "outs/per_sample_outs", x, "vdj_t/filtered_contig_annotations.csv"))
  tcr
})

cb <- combineTCR(tcrl, samples = samps, cells ="T-AB")

#check barcode prefixes
cb[[1]]$barcode
bcs

#we see that filtered_contig..csv has a _#_ prefix and our cds object has _#. prefix
#replace second underscore with a period in contig annotations

combined<- lapply(cb, function(x){
  bc<-str_split(x$barcode, "_")
  bcs<-sapply(bc, function(x){
    paste0(x[1], "_", x[2], ".", x[3])
  })
  x$barcode<-bcs
  x
})

#make sure the barcodes match!
combined[[1]]$barcode %in% colnames(cds) %>% table()
combined[[2]]$barcode %in% colnames(cds) %>% table()
combined[[3]]$barcode %in% colnames(cds) %>% table()

#add to seurat objects
seu4 <- combineExpression(combined, seu4, 
                  cloneCall="aa", 
                  proportion = FALSE, 
                  cloneTypes=c(Single=1, Small=5, Medium=20, Large=100, Hyperexpanded=500))
seu8 <- combineExpression(combined, seu8, 
                  cloneCall="aa", 
                  proportion = FALSE, 
                  cloneTypes=c(Single=1, Small=5, Medium=20, Large=100, Hyperexpanded=500))
```

# ID engineered T-cells in CD4 object
```{r ID TEa cells based on clonality & CD45.1}
FeaturePlot_scCustom(seu4, features = "Log_CD45_1", colors_use = paletteContinuous("whitePurple"))

df<-table(seu4$CTaa) %>% as.data.frame()
df<-df %>% arrange(desc(Freq))
clones<- df %>% slice_max(Freq, n =10)  
i<-1
for(i in 1:length(clones$Var1)){
  seu4 <- highlightClonotypes(seu4, 
                    cloneCall= "aa", 
                    sequence = clones$Var1[i] %>% as.character())

  pdf(paste0("../figs/cd4_vln_clone_cd45.1_", i, ".pdf"), width =7, height = 4)
  print(ggplot(seu4@meta.data, aes(x = highlight, y = log10(CD45_1)))+geom_violin())
  dev.off()
}


seu4 <- highlightClonotypes(seu4, 
                    cloneCall= "aa", 
                    sequence = clones$Var1[1] %>% as.character())


pdf("../figs/cd4_top_clone_umap.pdf", width = 7, height = 4)
DimPlot(seu4, group.by = "highlight", order = T)+scale_color_manual(values = c("#306300"), na.value = "gray80")&NoAxes()

seu4 <- highlightClonotypes(seu4, 
                    cloneCall= "aa", 
                    sequence = clones$Var1[4] %>% as.character())


pdf("../figs/cd4_top_clone_umap.pdf", width = 7, height = 4)
DimPlot(seu4, group.by = "highlight", order = T)+scale_color_manual(values = c("#306300"), na.value = "gray80")&NoAxes()
```

```{r cells with top clone are TEa cells given high CD45.1}
seu4$TEa<-seu4$highlight
seu4$TEa[seu4$highlight != "None"]<-"TEa"
table(seu4$TEa)
```

# assign mouse/celltype metadata
```{r}
df<-seu4@meta.data[, c("id", "seurat_clusters")]
seu4$mouse_cluster<-paste0(df[,1],"_cluster" ,df[,2])

df<-seu8@meta.data[, c("id", "sub.cluster")]
seu8$mouse_cluster<-paste0(df[,1],"_cluster" ,df[,2])
```

# Save object
```{r}
# saveRDS(cds, file.path(CDS_DIR, "ping_cds.rds"))
# saveRDS(seu, file.path(CDS_DIR, "ping_seu.rds"))
# saveRDS(seu4, file.path(CDS_DIR, "ping_seu4.rds"))
# saveRDS(cds4, file.path(CDS_DIR, "ping_cds4.rds"))
# saveRDS(seu8, file.path(CDS_DIR, "ping_seu8.rds"))
# saveRDS(cds8, file.path(CDS_DIR, "ping_cds8.rds"))
```
